### PR TITLE
Dynamically create blocks catalog

### PR DESCRIPTION
**Pull request by ahuang11**
Propagate work done in https://github.com/PrefectHQ/prefect-gcp/pull/96

### CODE REVIEW

File: {{cookiecutter.collection_name}}/docs/gen_ref_pages.py [hunk 1]
```diff
@@ -1,15 +1,104 @@
...
+        module_blocks[module_nesting].append(block_name)
+    return module_blocks
+
+
+def insert_blocks_catalog(generated_file):
+    module_blocks = find_module_blocks()
+    generated_file.write("## Blocks Catalog\n")
+    generated_file.write(
+        "Below is a list of Blocks available for registration in `{{ cookiecutter.collection_name }}`.\n\n"
+    )
```
Comment (ahuang11) [inline {{cookiecutter.collection_name}}/docs/gen_ref_pages.py:53]:
```suggestion
    generated_file.write(
        "Below is a list of Blocks available for registration in "
        "`{{ cookiecutter.collection_name }}`.\n\n"
    )
```

File: {{cookiecutter.collection_name}}/README.md [hunk 2]
```diff
@@ -38,13 +38,7 @@ Install `{{ cookiecutter.collection_name }}` with `pip`:
...
 ```
 
-Then, register to [view the block](https://orion-docs.prefect.io/ui/blocks/) on Prefect Cloud:
-
-```bash
-prefect block register -m {{ cookiecutter.collection_slug }}
-```
-
-Note, to use the `load` method on Blocks, you must already have a block document [saved through code](https://orion-docs.prefect.io/concepts/blocks/#saving-blocks) or [saved through the UI](https://orion-docs.prefect.io/ui/blocks/).
+A list of available blocks in `prefect-gcp` and their setup instructions can be found [here](https://github.com/{{ cookiecutter.github_organization }}/{{ cookiecutter.collection_name }}/#blocks-catalog/).
```
Comment (desertaxle) [inline {{cookiecutter.collection_name}}/README.md:41]:
```suggestion
A list of available blocks in `{{ cookiecutter.collection_name }}` and their setup instructions can be found [here](https://github.com/{{ cookiecutter.github_organization }}/{{ cookiecutter.collection_name }}/#blocks-catalog/).
```

File: {{cookiecutter.collection_name}}/docs/gen_ref_pages.py [hunk 3]
```diff
@@ -1,15 +1,105 @@
...
+        block_path = to_qualified_name(block)
+        if not block_path.startswith(COLLECTION_NAME):
+            continue
+        name_components = block_path.split(".")
+        module_nesting = tuple(name_components[1:-1])
+        block_name = name_components[-1]
+        if module_nesting not in module_blocks:
+            module_blocks[module_nesting] = []
+        module_blocks[module_nesting].append(block_name)
+    return module_blocks
```
Comment (desertaxle) [inline {{cookiecutter.collection_name}}/docs/gen_ref_pages.py:45]: Should we use the registry to find blocks like you did in https://github.com/PrefectHQ/prefect-collection-template/pull/114?

**Review by desertaxle — CHANGES_REQUESTED** [review]

File: {{cookiecutter.collection_name}}/docs/gen_ref_pages.py [hunk 4]
```diff
@@ -1,15 +1,94 @@
...
+                from prefect import flow
+                from {COLLECTION_SLUG}.{module_path} import {block_name}
+
+                @flow
+                def my_flow():
+                    my_block = {block_name}.load("MY_BLOCK_NAME")
+
+                my_flow()
+                ```
+                </details>
```
Comment (ahuang11) [inline {{cookiecutter.collection_name}}/docs/gen_ref_pages.py:78]:
```suggestion
```

File: {{cookiecutter.collection_name}}/docs/gen_ref_pages.py [hunk 5]
```diff
@@ -1,15 +1,93 @@
...
+
+
+def insert_blocks_catalog(generated_file):
+    module_blocks = find_module_blocks()
+    generated_file.write("## Blocks Catalog\n")
+    generated_file.write(
+        dedent(
+            f"""
+            Below is a list of Blocks available for registration in
+            `{{ cookiecutter.COLLECTION_SLUG }}`.
```
Comment (ahuang11) [inline {{cookiecutter.collection_name}}/docs/gen_ref_pages.py:40]:
```suggestion
            `{{ cookiecutter.collection_name }}`.
```

File: {{cookiecutter.collection_name}}/docs/gen_ref_pages.py [hunk 6]
```diff
@@ -1,15 +1,94 @@
...
+generates a list of them in the docs under the Blocks Catalog heading.
+"""
 
 from pathlib import Path
+from textwrap import dedent
 
 import mkdocs_gen_files
+from prefect.blocks.core import Block
+from prefect.utilities.importtools import to_qualified_name
+from prefect.utilities.dispatch import get_registry_for_type
```
Comment (desertaxle) [inline {{cookiecutter.collection_name}}/docs/gen_ref_pages.py:12]:
```suggestion
from prefect.utilities.dispatch import get_registry_for_type
from prefect.utilities.importtools import to_qualified_name
```

**Review by desertaxle — CHANGES_REQUESTED** [review]

**Review by desertaxle — APPROVED** [review]

**Pull request merged**